### ISSUE TITLE
needless_return doesn't trigger properly with tracing::instrument

### ISSUE THREAD

**Issue by joshuachp** [rust-lang/rust-clippy#13577]
### Summary

`needless_return` fails to trigger when the function has the macro attribute of `#[tracing::instrument]`. But it would trigger if the macro is expanded.

### Lint Name

needless_return

### Reproducer

Here is a link to the playground with the issue:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021

<details>

```rust
#[tracing::instrument]
fn foo() -> Result<Option<i32>, i32> {
    let res = Ok(42);
    match res {
        Ok(v) => Ok(Some(v)),
        Err(-1) => {
            return Err(-2);
        }
        Err(val) => return Err(val),
    }
}
```

### Version

rustc 1.82.0 (f6e511eec 2024-10-15)
binary: rustc
commit-hash: f6e511eec7342f59a25f7c0534f1dbea00d01b14
commit-date: 2024-10-15
host: x86_64-unknown-linux-gnu
release: 1.82.0
LLVM version: 19.1.1
